File: Week 1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 0);

    int length = 0;
    long temp = num;
    int sum = 0;
    bool isSecondDigit = false;

    while (temp != 0)
    {
        int digit = temp % 10;

        if (isSecondDigit)
        {
            digit *= 2;
            while (digit > 0)
            {
                sum += digit % 10;
                digit /= 10;
            }
        }
        else
        {
            sum += digit;
        }

        isSecondDigit = !isSecondDigit;
        temp /= 10;
        length++;
    }

    // Get the first two digits of the card number
    long firstTwoDigits = num;
    while (firstTwoDigits >= 100)
    {
        firstTwoDigits /= 10;
    }

    // Check card type
    if (sum % 10 == 0)
    {
        if ((length == 13 || length == 16) && (firstTwoDigits / 10 == 4))
        {
            printf("VISA\n");
        }
        else if (length == 15 && (firstTwoDigits == 34 || firstTwoDigits == 37))
        {
            printf("AMEX\n");
        }
        else if (length == 16 && (firstTwoDigits >= 51 && firstTwoDigits <= 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

File: Week 6/credit.py
def get_int():
    while True:
        try:
            return int(input("Number: "))
        except ValueError:
            pass


def check(num):
    digits = [int(digit) for digit in str(num)]
    count = 1
    sum1 = 0
    sum2 = 0
    for i in reversed(digits):
        if count % 2 == 0:
            doubled_digit = i * 2
            sum2 += doubled_digit // 10 + doubled_digit % 10
        else:
            sum1 += i
        count += 1
    total_sum = sum1 + sum2
    if len(digits) in [15, 16, 13]:
        if total_sum % 10 == 0:
            return True
    return False


x = get_int()
if check(x):
    if str(x)[:2] in ['34', '37']:
        print("AMEX")
    elif str(x)[0] == '4':
        print("VISA")
    elif 51 <= int(str(x)[:2]) <= 55:
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")